### DISCUSSION TITLE
Tutorial on setting up Searchable

### DISCUSSION THREAD

**Discussion by brianbiggs** [thoughtspot/cs_tools#128] [General]
Maybe I'm missing it, but I don't see any clear tutorial or instructions on how to use the Searchable functionality of CS Tools. I see the CS [Tools -> Tools -> Searchable](https://thoughtspot.github.io/cs_tools/tools/searchable/) page on github, and I am able to extract my metadata using `cs_tools tools searchable gather` command (BTW, the screenshots still show using the option `metadata` instead of `gather`).

But now what? I have the CSV files. Do I just have to `tsload` them into ThoughtSpot, build the joins between the tables, and build my own liveboards? I guess I was expecting to have all of that pre-built in a SpotApp so I don't have to reinvent the wheel.

I did see the `deploy` command to deploy the SpotApp, but when I tried to use that (command below), I got the error message **_The SpotApp is only configured for Connections at this time. Please reach out to the CS Tools team or your Solutions Consultant to get on the waitlist_**. Why are options for Falcon included if it's not supported? I feel like I'm just using it wrong.

`cstools tools searchable deploy --connection-guid falcon --database SEARCHABLE --schema falcon_default_schema --config CERT`

My client is in the process of moving from on-premise to the Cloud, so I could set up something in our cloud database environment, but I was just trying to set up something quick and dirty on-prem to get a picture of our configuration as we migrate to the ThoughtSpot Cloud.

Any help and hints are greatly appreciated!

Thanks,
Brian

**Comment by brianbiggs** [comment]
I kept digging, and found reference to use the Syncer to load it to ThoughtSpot Falcon. I tried the command below, but got errors. I am using 9.0.1.sw on-premise, CS Tools 1.4.15, and Python 3.11.8 on Windows 10. Maybe a compatibility issue with 9.0.1.sw?

`cstools tools searchable gather --config CERT --syncer falcon://database=SEARCHABLE`

```
    r = self.ts.api.dataservice_dataload_initialize(data=flags, timeout=http_timeout)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\cs_tools\api\_rest_api_v1.py", line 565, in dataservice_dataload_initialize
    fullpath = self.dataservice_url.copy_with(path="ts_dataservice/v1/public/loads")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\httpx\_urls.py", line 348, in copy_with
    return URL(self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\httpx\_urls.py", line 115, in __init__
    self._uri_reference = url._uri_reference.copy_with(**kwargs)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\httpx\_urlparse.py", line 134, in copy_with
    return urlparse("", **defaults)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\httpx\_urlparse.py", line 252, in urlparse
    validate_path(path, has_scheme=has_scheme, has_authority=has_authority)
  File "C:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\httpx\_urlparse.py", line 371, in validate_path
    raise InvalidURL("For absolute URLs, path must be empty or begin with '/'")
httpx.InvalidURL: For absolute URLs, path must be empty or begin with '/'
```

It created the tables in Falcon in the SEARCHABLE database, but no data is loaded.

**Reply by boonhapus** [reply]
You're actually on an older version of CS Tools -- we have 1.5.3 (or 1.5.4 depending on when you see this reply). I'd recommend upgrading, as my other answer (re: csv quoting) assumed you were on the 1.5.x series or releases.

The invalid url error here was a bug in 1.4 release which should be fixed. As well on 1.4, I had limited support for Falcon as the Searchable SpotApp was built using some SQL Views (which are not available in Falcon). We had a manual workaround which was available.

Generally the process here looks like..

1/ Run `searchable metadata --syncer falcon:// --config CERT` **
  - This will create the tables in Falcon @ the path `cs_tools.falcon_default_schema`
2/ Run `searchable deploy --export DIRECTORY --connection-guid falcon --database cs_tools --schema falcon_default_schema --config CERT`

_**If you were running Embrace/Connections, you'd want to run `searchable bi-server ...` to a connection type we support to mirror over TS: BI Server as well. Since you're on Falcon, we have a View that sits on top of TS: BI in order to support the SpotApp._

I would recommend you drop the `cs_tools` database that was created before running this above commands.

One thing to note is that the SpotApp was designed around Cloud, so there's potential that the Liveboard TML you export in 2/ will not directly import. The foundational layer of tables + views + worksheets should import however.

**Reply by boonhapus** [reply]
😅 Also the SpotApps that we have aren't yet well documented either.

**[CS Tools: Worksheet Column Utilization](https://thoughtspot.github.io/cs_tools/changelog/1.5.0/#__tabbed_1_2)**
**[CS Tools: TS BI Server Advanced](https://thoughtspot.github.io/cs_tools/changelog/1.5.0/#__tabbed_1_3)**

**Reply by brianbiggs** [reply]
I did a `cs_tools self upgrade` before I started on all this yesterday. It cranked out a bunch of messages (see below) that I interpreted as INFORMATIONAL and that the upgrade was fine. But now I see at the start it said it found 1.5.3 and I'm still on 1.4.15. Not sure why the self upgrade didn't work.

> [09:02:15] INFO     Getting the latest CS Tools release.                                                                                                           _self.py:68
           INFO     Found version: v1.5.3                                                                                                                          _self.py:70
           INFO     Upgrading CS Tools and its dependencies.                                                                                                       _self.py:77
[09:02:16] INFO     Collecting cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip (from cs_tools@                                             _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip)
[09:02:17] INFO     Downloading https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip                                                                      _updater.py:69
[09:02:19] INFO     \ 21.4 MB 24.6 MB/s 0:00:01                                                                                                                 _updater.py:69
[09:02:20] INFO     Installing build dependencies: started                                                                                                      _updater.py:69
[09:02:24] INFO     Installing build dependencies: finished with status 'done'                                                                                  _updater.py:69
           INFO     Getting requirements to build wheel: started                                                                                                _updater.py:69
[09:02:25] INFO     Getting requirements to build wheel: finished with status 'done'                                                                            _updater.py:69
           INFO     Installing backend dependencies: started                                                                                                    _updater.py:69
[09:02:27] INFO     Installing backend dependencies: finished with status 'done'                                                                                _updater.py:69
           INFO     Preparing metadata (pyproject.toml): started                                                                                                _updater.py:69
[09:02:28] INFO     Preparing metadata (pyproject.toml): finished with status 'done'                                                                            _updater.py:69
           INFO     Requirement already satisfied: pip>=23.1 in c:\users\bbiggs2\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (from cs_tools@           _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip->cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip) (24.0)
           INFO     Requirement already satisfied: thoughtspot-tml in c:\users\bbiggs2\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (from cs_tools@     _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip->cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip) (2.0.15)
[09:02:29] INFO     Requirement already satisfied: awesomeversion in c:\users\bbiggs2\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (from cs_tools@      _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip->cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip) (24.2.0)
           INFO     Requirement already satisfied: httpx>=0.27.0 in c:\users\bbiggs2\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (from cs_tools@       _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip->cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip) (0.27.0)
           INFO     Requirement already satisfied: pydantic>=2.6.4 in c:\users\bbiggs2\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (from cs_tools@     _updater.py:69
                    https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip->cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.5.3.zip) (2.7.1)

**Reply by brianbiggs** [reply]
I also found [this link about re-installing](https://github.com/thoughtspot/cs_tools/discussions/117#discussioncomment-8990015). I tried that, and it appeared to install 1.5.3 again. However, I'm still showing that I'm on version 1.4.15. 😠 

Digging into that a little more, somehow I had an old copy of cs_tools.exe and cstools.exe (1.4.15) in **_C:\Users\User\AppData\Local\Programs\Python\Python311\Scripts_**. Not sure how it ended up in there, but that was getting picked up before the updated copy in **_C:\Users\bbiggs2\AppData\Roaming\cs_tools\.cs_tools\Scripts_**. 🤷‍♂️ 

It's now picking up the newer version, so I'll follow your instructions above and see if I can get this working.

**Reply by brianbiggs** [reply]
Now I'm getting this error. I believe I have access to our URL on that port.

```
[16:34:28] INFO     Attempting Basic authentication                                                                                                         thoughtspot.py:213
           INFO     Initializing syncer: <class 'cs_tools_falcon_syncer.Falcon'>                                                                                  syncer.py:50
                                                                           Searchable Data Gatherer

                                                   Org        User        Group        Tag       Object      Column     Dependent    Access
                                ──────────────────────────────────────────────────────────────────────────────────────────────────────────────
                                  ThoughtSpot                  📁          📁          📁          📁          📁          📁          📁

                                                                                fetching data
┌──────────────────────────────────────────────────────────────────── The tsload service is unreachable. ────────────────────────────────────────────────────────────────────┐
│ could not connect at https://<our_ts_url>:8442                                                                                                                             │
│                                                                                                                                                                            │
│ Mitigation                                                                                                                                                                 │
│ Is your VPN connected?                                                                                                                                                     │
│                                                                                                                                                                            │
│ If that isn't the URL of your ThoughtSpot cluster, then your ThoughtSpot admin likely has configured the Remote TSLoad Connector Service to use a load balancer and your   │
│ local machine is unable to connect directly to the ThoughtSpot node which is accepting files.                                                                              │
│                                                                                                                                                                            │
│ You can try using `ignore_load_balancer_redirect` in your Falcon syncer definition as well.                                                                                │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

I tried this:
`cstools tools searchable metadata --config CERT --syncer falcon://ignore_load_balance_redirect=true`
and this:
`cstools tools searchable metadata --config CERT --syncer falcon://ignore_load_balance_redirect=1`
Not sure of the correct format I should be using for that parameter, but neither worked. It just runs for a couple minutes and looks like it eventually times out with that same `tsload service is unreachable` error.

**Reply by boonhapus** [reply]
@brianbiggs are able to ping the dataload service at all?

`curl -i https://<our_ts_url>:8442/ts_dataservice/v1/public/ping`

If the service (`etl_http_server`) isn't responding or hasn't been enabled, you might need to ~turn it on~ restart it.

**Reply by brianbiggs** [reply]
@boonhapus, I get this message **_immediately_** when I try to ping either our non-prod or prod on-prem cluster from my laptop.

```
$ curl -i https://<out_ts_surl>:8442/ts_dataservice/v1/public/ping
curl: (35) schannel: failed to receive handshake, SSL/TLS connection failed
```

I made sure that my laptop has firewall access to those servers on port 8442. The `etl_http_server` has always been running, but I did restart it as well, just to make sure.

When I try the same `curl` command from the non-prod (or prod) cluster itself (logged in as admin), I get the message below after about a minute. I would think that being directly connected to the cluster would get a response.

```
[admin@<hostname> ~]$ curl -i https://<our_ts_url>:8442/ts_dataservice/v1/public/ping
curl: (7) Failed connect to <our_ts_url>:8442; Connection timed out
```

Maybe listening on a different port? Is this the correct API location? Something different about ThoughtSpot 9.0.1.sw? Other thoughts?

Should this same `curl` command work on our ThoughtSpot Cloud clusters, because I get a timeout when trying to ping those from my laptop as well.

```
$ curl -i https://<cloudhost>.thoughtspot.cloud:8442/ts_dataservice/v1/public/ping
curl: (28) Failed to connect to <cloudhost>.thoughtspot.cloud port 8442 after 43200 ms: Couldn't connect to server
```

**Reply by boonhapus** [reply]
> I get this message immediately when I try to ping either our non-prod or prod on-prem cluster from my laptop.

This sounds like a network issue to me @brianbiggs . ThoughtSpot won't respond so quickly with an ACK like this, but a TLS preflight check would.

> Maybe listening on a different port?

This is possible, but unlikely IMO. If you tscli list overrides for `etl_http_server.etl_http_server` , it should show you the port number for the service.

FWW 🧙 There's no magic here. ✨ This is the API we're interacting with: https://docs.thoughtspot.com/software/latest/tsload-connector

> Should this same curl command work on our ThoughtSpot Cloud clusters ...
No, the service behind the tsload connector is called `etl_http_server` , and it's only available on Software clusters (where Falcon is available). 

I'm very surprised to see you get a timeout when you're on local. Is your `<our_ts_url>` a domain which is actually pointed to a load balancer (and then on to a bundle of nodes)? If so, that might be why you're getting a failed to connect, the load balancer might not handle traffic on any port other than `:8080` or `:8000` or similar. You could try switching the URL to an IP of one of the nodes in the cluster itself.

**Comment by boonhapus** [comment]
You're totally right about it not being spelled out in the docs @brianbiggs . I've got a few open tasks to improve the documentation (both shortening the  Getting Started , providing some Guides for some of the tools that need them, and adding an FAQ for some common question -- like your csv one in #129 ). I'd love to hear your suggestions for these!

These could be coming out over the next month or so to make the documentation a bit more friendly.

**Comment by boonhapus** [comment]
>My client is in the process of moving from on-premise to the Cloud, so I could set up something in our cloud database environment, but I was just trying to set up something quick and dirty on-prem to get a picture of our configuration as we migrate to the ThoughtSpot Cloud.

One thing that we do for Pro-Serve supported migrations is use Searchable exported to sqlite, and build some reports on top of that. I've built a separate utility to help run migrations that comes along with the services engagement (ie, it's not public 😅), but we leverage Searchable's output to do some pre-migration cleanup and exploration prior to scoping out the full project.

```
cs_tools tools searchable metadata --syncer sqlite://database_path=...
```

The **[data model that's in the docs](https://thoughtspot.github.io/cs_tools/tools/searchable/)** reflects that core model and joins, and I'm currently in the process of cleaning this up to include the new tables added to Searchable in 1.5 (which wouldn't exist in SW 9.0.1 , as they're centered around Orgs)

![](https://thoughtspot.github.io/cs_tools/tools/searchable/searchable-erd.png)
